### PR TITLE
Syntax highlighting for folder-level PR diffs

### PR DESCRIPTION
**Pull request by alejandro5042**
Uses [highlight.js](https://highlightjs.org/) and the GitHub language definitions to support 189 languages with 1,167 extensions (e.g. `.csproj` is an XML file) and 256 special filenames (e.g. `bashrc` is an sh file).

### CODE REVIEW

File: src/azdo-pr-dashboard.user.js [hunk 1]
```diff
@@ -930,6 +936,221 @@
...
+          span.innerText = fragment.textContent;
+          fragment.parentNode.replaceChild(span, fragment);
+        }
+      }
+    }
+  }
+
+  // Helper function to get the file extension out of a file path; e.g. `cs` from `blah.cs`.
+  function getFileExt(path) {
+    return /(?:\.([^.]+))?$/.exec(path)[1];
```
Comment (m-akinc) [inline src/azdo-pr-dashboard.user.js:1151]: This works? Doesn't the capturing group (that matches the actual extension) need to be outside the non-capturing group (that matches the '.')?

File: src/azdo-pr-dashboard.user.js [hunk 2]
```diff
@@ -930,6 +936,221 @@
...
+        width: 100%;
+    }`);
+
+  function watchForNewDiffs() {
+    eus.onUrl(/\/pullrequest\//gi, (session, urlMatch) => {
+      let languageDefinitions = null;
+      session.onEveryNew(document, '.text-diff-container', diff => {
+        if (eus.seen(diff)) return;
+
+        // Only parse languages if we have something to diff.
```
Comment (m-akinc) [inline src/azdo-pr-dashboard.user.js:1056]: This comment seems unrelated to the code around it.

File: src/azdo-pr-dashboard.user.js [hunk 3]
```diff
@@ -930,6 +936,221 @@
...
+          const extension = getFileExt(fileName);
+
+          const leftPane = diff.querySelector('.leftPane > div > .side-by-side-diff-container');
+          const rightOrUnifiedPane = diff.querySelector('.rightPane > div > .side-by-side-diff-container') || diff;
+
+          // Guess our language based on our file extension.
+          // Supports languages listed here, without plugins: https://github.com/highlightjs/highlight.js/blob/master/SUPPORTED_LANGUAGES.md
+          let language = null;
+          for (const mode of [extension].concat(languageDefinitions.extensionToMode[extension]).concat(languageDefinitions.fileToMode[fileName])) {
+            if (hljs.getLanguage(mode)) {
```
Comment (m-akinc) [inline src/azdo-pr-dashboard.user.js:1074]: Is this check necessary? Seems like the maps should only be populated with modes for supported languages.

Comment (alejandro5042) [reply src/azdo-pr-dashboard.user.js:1074]: Ideally yes, however, the GitHub language keywords and the highlight.js language keywords are not the same. It's two different projects. They do often match, so that's why I'm in this loop trying to find a match within the array. I can document it a bit more. It's a heuristic to find a language match.

File: src/azdo-pr-dashboard.user.js (hunk 2 above)
Comment (alejandro5042) [reply src/azdo-pr-dashboard.user.js:1056]: It makes sense to me, but yes agreed -- very poorly worded :) Parsing the language definitions takes CPU time, so we only want to do it if we have a diff container to try to highlight. But the comment doesn't really add anything so I'll remove it.

File: src/azdo-pr-dashboard.user.js (hunk 1 above)
Comment (alejandro5042) [reply src/azdo-pr-dashboard.user.js:1151]: `?:` represents a non-capturing group but internally it has a capturing group, which is the only one in the regex, hence we can find it at index 1. It does work since it's been highlighting my files for the past few works :)
